contracts/btc_swap/lib.rs

```diff
@@ -0,0 +1,145 @@
+        transaction: Transaction,
+        length_bound: u32,
+        address: Vec<u8>,
+    ) -> Option<u64>;
+}
+
+#[derive(Debug, Copy, Clone, PartialEq, Eq, scale::Encode, scale::Decode)]
+#[cfg_attr(feature = "std", derive(scale_info::TypeInfo))]
+pub enum RuntimeErr {
+    SomeFailure,
+}
+
+impl ink::env::chain_extension::FromStatusCode for RuntimeErr {
+    fn from_status_code(status_code: u32) -> Result<(), Self> {
+        match status_code {
+            0 => Ok(()),
+            1 => Err(Self::SomeFailure),
+            _ => panic!("encountered unknown status code"),
+        }
+    }
+}
+
+/// Creates a swap contract where Alice locks DOT with a price limit in a contract that
+/// Bob can a acquire by sending BTC on the Bitcoin chain that Alice provided.
+///
+/// Note: this is a proof of concept protocol and should not be used in production due to flaws in
+/// the protocol and implementation.
+///
+/// ## Protocol
+///
+/// - Alice provides a BTC address, a price limit, and a DOT amount to lock in the contract.
+/// - Bob sends BTC to the address provided by Alice.
+/// - Anyone (Alice, Bob, or a third party) provides a BTC transaction proof to the contract.
+/// The proof triggers a DOT transfer from the contract to Bob.
+#[ink::contract(env = crate::CustomEnvironment)]
+mod btc_swap {
+    use super::*;
+    use bitcoin::Address as BtcAddress;
+    use ink::storage::Mapping;
+    use scale::Encode;
+
+    /// Defines the limit order contract
+    #[derive(scale::Decode, scale::Encode)]
+    #[cfg_attr(
+        feature = "std",
+        derive(Debug, PartialEq, Eq, scale_info::TypeInfo, ink::storage::traits::StorageLayout)
+    )]
+    pub struct LimitOrder {
+        /// The BTC address to send BTC to
+        /// can't store as `BtcAddress`, since `StorageLayout` is not implemented, and we can't derive it
+        /// due to the hashes inside it not implementing `StorageLayout`
+        btc_address: Vec<u8>,
+        /// The price limit for the BTC denoted in satoshis
+        min_satoshis: u64,
+        /// The DOT amount to lock in the contract denoted in planck
+        plancks: u128,
+    }
+
+    /// Defines the storage of your contract.
+    /// Add new fields to the below struct in order
+    /// to add new static storage fields to your contract.
+    #[ink(storage)]
+    pub struct BtcSwap {
+        /// Stores a mapping from an account to a limit order.
+        orders: ink::storage::Mapping<AccountId, LimitOrder>,
+    }
+
+    impl BtcSwap {
+        #[ink(constructor)]
+        pub fn new() -> Self {
+            let orders = Mapping::default();
+            Self { orders }
+        }
+
+        #[ink(message, payable)]
+        pub fn create_trade(&mut self, btc_address: BtcAddress, min_satoshis: u64) {
+            assert!(min_satoshis > 0);
+
+            let caller = self.env().caller();
+            let offer = self.env().transferred_value();
+
+            let order = LimitOrder {
```

Comment:
I realize this is only an example but ideally we would include and check an `OP_RETURN` value in this contract, should we expose some way to generate that through the chain extensions?